CenterForOpenScience/osf.io
# [Migration Prep Needed] Feature/Meetings v2 Optimizations [ENG-455]

❗️ ~~Solution needs Product Approval~~ Approved

## Purpose

There are two issues with the current meetings v2 implementation at scale:

### Issue 1
Private `MeetingsList` and `MeetingsDetail` endpoints differ on submission counts. (Detail is correct).  List endpoint annotates queryset with submission count for sorting purposes and detail endpoint calculates on the fly.  Cause is case-sensitivity issue - tags on meetings with different casing are still counted as submissions to the meeting.  We're building a weird subquery to try to deal with the fact that nodes aren't actually related to conferences, but this complication makes things difficult.

### Issue 2
MeetingSubmissionList endpoint timing out.   Slow part is annotating queryset with download counts for sorting purposes -  this requires a sequence scan of the PageCounter table, because PageCounters are not actually related to Files, we're just checking for string matching.

## Changes

### For Issue 1
Added a `submissions` m2m field to the Conference model to actually relate Conferences and AbstractNodes in the db. The queries are getting too cumbersome in Django as these items were not actually related.
- Added a schema/data migration to add all submissions with matching tags to each conference
- The email process to create a conference submission relates the node to the conference rather than adding a tag
- Gets rid of cached num_submissions field - v1 was writing to this field but no longer needed
- Updates legacy v1 page to accommodate new conference-submission relationship until new emberized page is waffled on
- Now annotating queryset with submissions_count is easy (and possible)

### For Issue 2 
- Temporarily disables sorting on download count until PageCounter work gets in. 
- Improved how download count is calculated on the fly 
- The combination causes acceptable speeds and PageCounter improvements + enabling sorting will also have acceptable speeds.

## QA Notes
😞 
- Meetings needs another full regression of both legacy page and emberized page, as both have now been modified. 
- The email process to add a submission to a Conference needs to be tested.
- Adding a meeting tag does *not* add a project to a Conference - current workflow is now only through email (though I can add this somewhere else if need be)
- Verify submissions count on MeetingsList matches the MeetingsDetail page (before migration -add projects to meetings by adding tags with different casing - `spsp2016` versus `SPSP2016`

## Documentation

Solution involves temporarily disabling functionality, but new queries to replace are documented here.

## Side Effects

- No place in UI for user to remove a project from a conference other than marking the project private.
  - We can come up with a place to surface this, or if it's not a common action, can add to the admin app for QA.
- Sorting on download counts for submissions is temporarily disabled until we get in [PageCounter improvements](https://openscience.atlassian.net/browse/ENG-122) 

## Ticket

https://openscience.atlassian.net/browse/ENG-455
